lib/appium.js

```diff
@@ -165,37 +175,47 @@ class AppiumDriver extends BaseDriver {
 
     // sessionOverride server flag check
     // this will need to be re-thought when we go to multiple session support
-    if (this.args.sessionOverride && !!this.sessions && _.keys(this.sessions).length > 0) {
-      log.info('Session override is on. Deleting other sessions.');
-      for (let id of _.keys(this.sessions)) {
-        log.info(`    Deleting session '${id}'`);
-        try {
-          await this.deleteSession(id);
-        } catch (ign) {
-          // the error has already been logged in AppiumDriver.deleteSession
-          // continue
+    if (this.args.sessionOverride) {
+      const sessionIdsToDelete = await sessionsListGuard.acquire(AppiumDriver.name, () => _.keys(this.sessions));
+      if (sessionIdsToDelete.length) {
+        log.info(`Session override is on. Deleting other ${sessionIdsToDelete.length} active session${sessionIdsToDelete.length ? '' : 's'}.`);
+        for (const id of sessionIdsToDelete) {
+          log.info(`\tDeleting session '${id}'`);
+          try {
+            await this.deleteSession(id);
+          } catch (ign) {
+            // the error has already been logged in AppiumDriver.deleteSession
+            // continue
+          }
         }
       }
     }
 
-    let runningDriversData;
+    let runningDriversData, otherPendingDriversData;
+    let d = new InnerDriver(this.args);
     try {
-      runningDriversData = this.curSessionDataForDriver(InnerDriver);
+      runningDriversData = await this.curSessionDataForDriver(InnerDriver);
     } catch (e) {
       throw new errors.SessionNotCreatedError(e.message);
     }
-
-    let innerSessionId, dCaps;
-    let d = new InnerDriver(this.args);
-    try {
+    await pendindDriversGuard.acquire(AppiumDriver.name, () => {
       this.pendingDrivers[InnerDriver.name] = this.pendingDrivers[InnerDriver.name] || [];
-      const otherPendingDriversData = this.pendingDrivers[InnerDriver.name].map((drv) => drv.driverData);
+      otherPendingDriversData = this.pendingDrivers[InnerDriver.name].map((drv) => drv.driverData);
       this.pendingDrivers[InnerDriver.name].push(d);
+    });
+    let innerSessionId, dCaps;
+    try {
       [innerSessionId, dCaps] = await d.createSession(caps, reqCaps, [...runningDriversData, ...otherPendingDriversData]);
     } finally {
-      _.pull(this.pendingDrivers[InnerDriver.name], d);
+      await pendindDriversGuard.acquire(AppiumDriver.name, () => {
+        _.pull(this.pendingDrivers[InnerDriver.name], d);
+      });
+      await sessionsListGuard.acquire(AppiumDriver.name, () => {
+        if (innerSessionId) {
```

Comment:
actually now that i'm looking at this, why wouldn't we want this to be inside the `try` block? then we don't need an `if` statement at all because we know we have a session.